Bug report and bugfix for problem with invalid characters in thrown exception

slavo2 opened the pull request:
#### Bug description:

KlarnaException which can be attempted to be thrown at line of 3669 of Klarna.php might contain possibly invalid characters from incorrectly interpreted faultString().
#### Bug fix:

Converting the character encoding of the message before being used in constructing KlarnaException by adding and using method Klarna::fixMessageEncoding()

Commit:
Klarna.php : adding and using method fixMessageEncoding()
 - fix for problem with (possibly) invalid characters in thrown exception

JoakimLofgren commented:
Thanks for your PR, I will take a look at it.

JoakimLofgren commented on Klarna.php:
```diff
@@ -4417,6 +4418,19 @@ private function _checkConfig($config = null)
+     * @param string $originalMessage
+     * @return string
+     */
+    private function fixMessageEncoding($originalMessage)
```
_private_ methods should be prefixed with an underscore.

JoakimLofgren commented on Klarna.php:
```diff
@@ -4417,6 +4418,19 @@ private function _checkConfig($config = null)
+    private function fixMessageEncoding($originalMessage)
+    {
+        $possibleInputEncodings = ['ISO-8859-1', 'UTF-8'];
+        $outputEncoding = mb_internal_encoding();
```
> mbstring is a non-default extension. This means it is not enabled by default. 

Probably need something like:

``` php
if (!extension_loaded('mbstring')) {
    return $originalMessage;
}

...
```

As encodings in PHP is a bit messy, I don't know if using `iconv` or the `utf8_decode`/`utf8_encode` might be a more generic solution that everyone can benefit from.

slavo2 committed:
Private method Klarna::fixMessageEncoding() renamed to _fixMessageEncoding() and updated:
support for intl, mbstring and iconv php extensions
fallback to ASCII using regular expression
possibility to specify input and output encoding as method arguments

slavo2 committed:
Private method Klarna::_fixMessageEncoding() update, refactoring and phpDoc update

slavo2 commented:
I have renamed and updated the private method (now:) Klarna::_fixMessageEncoding()

JoakimLofgren commented:
We're working on getting this repo into composer. As soon as that is done, I'll look into getting your stuff merged.

slavo2 commented:
Hi Joakim

I have already made a composer package from your SDK
https://github.com/Subscribo/klarna-invoice-sdk-wrapped
feel free to use (anything from) it.

Slavo

slavo2 commented:
Alternative (and possibly better) solution provided in pull request https://github.com/klarna/php-xmlrpc/pull/3
